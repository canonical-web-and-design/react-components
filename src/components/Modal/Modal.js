import { nanoid } from "nanoid";
import PropTypes from "prop-types";
import React, { useRef } from "react";

export const Modal = ({ children, close, title }) => {
  const descriptionId = useRef(nanoid());
  const titleId = useRef(nanoid());
  return (
    <div className="p-modal" onClick={close}>
      <div
        className="p-modal__dialog"
        role="dialog"
        aria-labelledby={titleId.current}
        aria-describedby={descriptionId.current}
        onClick={(evt) => evt.stopPropagation()}
      >
        <header className="p-modal__header">
          <h2 className="p-modal__title" id={titleId.current}>
            {title}
          </h2>
          <button
            className="p-modal__close"
            aria-label="Close active modal"
            onClick={close}
          >
            Close
          </button>
        </header>
        <div id={descriptionId.current}>{children}</div>
      </div>
    </div>
  );
};

Modal.propTypes = {
  children: PropTypes.node.isRequired,
  close: PropTypes.func.isRequired,
  title: PropTypes.string.isRequired,
};

export default Modal;
