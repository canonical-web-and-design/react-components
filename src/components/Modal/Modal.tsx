import PropTypes from "prop-types";
import React, { useRef } from "react";
import type { ReactNode } from "react";
import uuidv4 from "uuid/v4";

type Props = {
  children: ReactNode;
  close: () => void;
  title: string;
};

export const Modal = ({ children, close, title }: Props): JSX.Element => {
  const descriptionId = useRef(uuidv4());
  const titleId = useRef(uuidv4());
  return (
    <div className="p-modal" onClick={close}>
      <div
        className="p-modal__dialog"
        role="dialog"
        aria-labelledby={titleId.current}
        aria-describedby={descriptionId.current}
        onClick={(evt) => evt.stopPropagation()}
      >
        <header className="p-modal__header">
          <h2 className="p-modal__title" id={titleId.current}>
            {title}
          </h2>
          <button
            className="p-modal__close"
            aria-label="Close active modal"
            onClick={close}
          >
            Close
          </button>
        </header>
        <div id={descriptionId.current}>{children}</div>
      </div>
    </div>
  );
};

Modal.propTypes = {
  children: PropTypes.node.isRequired,
  close: PropTypes.func.isRequired,
  title: PropTypes.string.isRequired,
};

export default Modal;
