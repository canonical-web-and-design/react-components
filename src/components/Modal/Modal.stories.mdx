import { ArgsTable, Canvas, Meta, Story } from "@storybook/addon-docs/blocks";

import Modal from "./Modal";

<Meta
  title="Modal"
  component={Modal}
  argTypes={{
    children: {
      control: {
        disable: true,
      },
    },
  }}
/>

export const Template = (args) => {
  return (
    <div style={{ height: "100vh", position: "relative" }}>
      <Modal {...args} />
    </div>
  );
};

### Modal

This is a [React](https://reactjs.org/) component for the Vanilla [Modal](https://docs.vanillaframework.io/patterns/modal/).

The modal component can be used to overlay an area of the screen which can contain a prompt, dialog or interaction.

### Props

<ArgsTable of={Modal} />

<Canvas>
  <Story
    name="Default"
    args={{
      close: () => {},
      title: "Confirm delete",
      children: (
        <>
          <p>
            Are you sure you want to delete user "Simon"? This action is
            permanent and can not be undone.
          </p>
          <hr />
          <div className="u-align--right">
            <button onClick={() => {}} className="u-no-margin--bottom">
              Cancel
            </button>
            <button className="p-button--negative u-no-margin--bottom">
              Delete
            </button>
          </div>
        </>
      ),
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>
