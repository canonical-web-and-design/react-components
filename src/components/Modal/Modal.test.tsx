import { shallow } from "enzyme";
import React from "react";

import Modal from "./Modal";

jest.mock("uuid/v4", () =>
  jest.fn(() => "00000000-0000-0000-0000-000000000000")
);

describe("Modal ", () => {
  it("renders", () => {
    const wrapper = shallow(
      <Modal close={jest.fn()} title="Confirm delete">
        Are you sure?
      </Modal>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
