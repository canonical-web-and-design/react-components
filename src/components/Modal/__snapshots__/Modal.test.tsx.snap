// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal  renders 1`] = `
<div
  className="p-modal"
  onClick={[MockFunction]}
>
  <div
    aria-describedby="00000000-0000-0000-0000-000000000000"
    aria-labelledby="00000000-0000-0000-0000-000000000000"
    className="p-modal__dialog"
    onClick={[Function]}
    role="dialog"
  >
    <header
      className="p-modal__header"
    >
      <h2
        className="p-modal__title"
        id="00000000-0000-0000-0000-000000000000"
      >
        Confirm delete
      </h2>
      <button
        aria-label="Close active modal"
        className="p-modal__close"
        onClick={[MockFunction]}
      >
        Close
      </button>
    </header>
    <div
      id="00000000-0000-0000-0000-000000000000"
    >
      Are you sure?
    </div>
  </div>
</div>
`;
